src/lib/virtio-net/src/hal.rs

```diff
@@ -0,0 +1,193 @@
+                }
+            }
+        },
+        BufferDirection::Both => match DMA_POOL_BIDIRECTIONAL.get() {
+            Some(pool) => pool,
+            None => {
+                let pool = DmaPool::new(
+                    DmaPool::default_spec(),
+                    Access::BiDirectional,
+                    DmaOptions::empty(),
+                );
+                match DMA_POOL_BIDIRECTIONAL.set(pool) {
+                    Ok(_) => {}
+                    Err(_) => panic!("Failed to set DMA_POOL_BIDIRECTIONAL"),
+                };
+                match DMA_POOL_BIDIRECTIONAL.get() {
+                    Some(pool) => pool,
+                    None => panic!("Failed to set DMA_POOL_BIDIRECTIONAL"),
+                }
+            }
+        },
+    }
+}
+
+fn insert_alloced(paddr: PhysAddr, dma_slice: DmaSliceRegion<u8>) {
+    let dict = ALLOCED.get_or_init(|| Mutex::new(HashMap::new()));
+
+    let wrapped = Fragile::new(dma_slice);
+
+    dict.lock().unwrap().insert(paddr, wrapped);
+}
+
+fn remove_alloced(paddr: PhysAddr) -> Option<Fragile<DmaSliceRegion<u8>>> {
+    let dict = ALLOCED.get_or_init(|| Mutex::new(HashMap::new()));
+
+    dict.lock().unwrap().remove(&paddr)
+}
+
+unsafe impl Hal for TestHal {
+    // Required methods
+    fn dma_alloc(pages: usize, direction: BufferDirection) -> (PhysAddr, NonNull<u8>) {
+        assert!(pages == 1, "Only 1 page supported");
+
+        let pool = get_dma_pool(direction);
+        let alloced = pool.allocate_array(pages, 0u8).unwrap();
+
+        let mut dma_slice = alloced;
+
+        let pin = dma_slice.pin().unwrap();
+        let phys_addr: virtio_drivers::PhysAddr =
+            u64::from(pin.into_iter().next().unwrap().addr()) as virtio_drivers::PhysAddr;
+        let virt = unsafe { NonNull::<u8>::new(dma_slice.get_mut().as_mut_ptr()) }.unwrap();
+
+        // Persist the allocated memory so it isn't freed when the function returns
+        insert_alloced(phys_addr, dma_slice);
+
+        // println!("Allocated DMA buffer at: {:?} with phys addr: {:x}", virt, phys_addr);
+
+        (phys_addr as PhysAddr, virt)
+    }
+
+    unsafe fn dma_dealloc(_paddr: PhysAddr, _vaddr: NonNull<u8>, _pages: usize) -> i32 {
+        let dma_region = remove_alloced(_paddr).unwrap().try_into_inner();
+        match dma_region {
+            Ok(mut dma_region) => {
+                dma_region.release_pin();
+                0
+            }
+            Err(_) => 1,
+        }
+    }
+
+    unsafe fn mmio_phys_to_virt(_paddr: PhysAddr, _size: usize) -> NonNull<u8> {
+        panic!("Should never be called as we have our own transport implementation");
+    }
+
+    unsafe fn share(buffer: NonNull<[u8]>, direction: BufferDirection) -> PhysAddr {
+        // println!("Sharing buffer at: {:?}", buffer);
+        let buf_len = buffer.len();
+
+        assert!(buf_len <= DMA_PAGE_SIZE, "Hal::Share(): Buffer too large");
+        let (phys, virt) = TestHal::dma_alloc(1, direction);
+
```

Comment:
I'd pull out a lot of this white space.  I'm from the old days where we only had 24 lines on the screen at once ;-)

Reply:
Took out about 20 lines of whitespace in this this file, any other particularly bad offenders I should fix?